Update master plan (#647)
sourcegraph/about: company/strategy.md

# Sourcegraph master plan

Everything we do stems from our [purpose](#purpose), to help *everyone* build software. We'll accomplish that over the next 30 years by following our strategy:

- 30 years
  - [Purpose](#purpose)
  - [Mission](#mission)
  - [Values](#values)
- 10 years
  - [10-year vision](#10-year-vision)
  - [Big Code](#big-code)
- 3 years
  - [3-year vision](#3-year-vision)
  - [Principles](#principles)
  - [Assumptions](#assumptions)
- 1 year
  - [1-year vision](#1-year-vision)
  - [Product direction](#product-direction)
  - [OKRs](#okrs)

## Purpose

> We want everyone to be able to build software. A world where everyone, not just the ~0.2% who can code today, builds software will see faster and more broadly beneficial technological progress.

Reviewer: I know this has been up for awhile, but is this really our purpose? What immediately comes to mind when reading this purpose is some sort of learning tool that teaches non-developers the fundamentals of programming. That is not what we do. What we do is understand the universe of code and put that at every developer's fingertips. This is useful to both experienced developers and newbies, but doesn't feel the same as enabling "everyone to be able to build software".

Alternatives:
* We want everyone to be able to participate in the building of software.
* We want to make software development effective at scale.
* More and more people are going to build software, and Sourcegraph is the tool the world needs to handle the influx of new developers and code.
* We want to make the universe of code accessible to every developer, regardless of where they are in the world, what company they work for, and their experience level.

Reviewer 2: I don't think our goal is to help everyone write software, for the same reasons Beyang mentions. But there is a need for many people to understand complex and large codebases. What is the system of record for VP Engineerings?

### 10-year vision {#10-year-vision}

> Problem statement: Software development is a black box to everyone else in an organization. Communication and collaboration in both directions is poor.

Everyone involved in the software value chain (developers, sales, product, marketing, support, customers/users, etc.) uses Sourcegraph to coordinate work and share information.

From a developer's point of view, they can go from the code for a particular feature to a list of all customers using it, all feedback and support requests about it, and all upcoming features that relate to it.

From a customer's and salesperson's point of view, they can see all in-progress or planned features for the particular customer. For any given feature, they can see the current state of development.

> Unlike the [3-year vision](#3-year-vision), our 10-year vision involves more kinds of people using Sourcegraph (not just people who write code directly).

## Principles

- Sourcegraph is universal code search, not universal "everything" search. Any additional data types in our search need to be relevant to the software development workflow.

> - We're a mass-market company and must maintain broad appeal. We want every developer, not just a specific niche audience, to use Sourcegraph.

Reviewer: This strikes me as more of a desired outcome, rather than a principle. "Mass-market" also sounds kinda negative (mass-market == commodity == cheap == primarily concerned about cash).

In my view, a more principle-y statement would be: "We wish to serve developers across a broad spectrum of communities, styles, and backgrounds. We are un-opinionated when it comes to specific languages and technologies, as we recognize that innovation comes from across the spectrum. We trust our customers to choose the best technologies for the job, and seek to make them productive in whatever stack they choose."

Author: Updated to just say:

```
- We want every developer, not just a specific niche audience, to use Sourcegraph.
  - We won't adopt polarizing public stances that would divide our audience.
```

Reviewer 3: >  - We won't adopt polarizing public stances that would divide our audience.

If this is intended to refer to technical stances (languages, frameworks, etc.)? Just from an outsider's POV, it sounds more broad than that the way it's phrased.

Author: I intended it to be about technical stances, but it would also apply to other stances. Of course, it's a judgment call as to whether a particular stance would "divide" our audience or appeal to it.

  - Any given developer will only pick one code search tool to use. Any given company will standardize on a single code search tool. Therefore, to avoid fragmentation, Sourcegraph should be not only *much* better than the alternatives, but also *not worse* in any significant way.
  - We won't adopt polarizing public stances that would divide our audience.
- Our product should strive to be fundamentally privacy-respecting and secure. This means that users don't need to trust us to verify that their data is private and secure.
- Build on-ramps in our product to turn more people into frequent users, instead of building the product *for* infrequent users (which is a self-fulfilling prophecy).
- We eventually want to be a platform that ties together all of the tools developers use.

>   - Other developer tools are partners, not as competitors.

Reviewer: ```suggestion
  - Other developer tools are partners, not competitors.
```

>   - This entails designing more extensibility into our product (and documenting it more thoroughly) than what's needed today, in order to make it easy for other people to build on.

Reviewer: ```suggestion
  - This entails designing for extensibility in our product (and documenting it thoroughly) as a first-order priority.
```

Author: updated

> 

Reviewer: Additional principle:
- Sourcegraph provides greater value the larger a software team. We are building not just for individual developers, but for development teams. We believe software development is increasingly a multiplayer game.

## Assumptions

- Sufficiently good code search will be useful to every developer many times per day (on average). It may take a while to convert any specific person into a frequent code search power user, but it will happen eventually.

> - Code search that is *exclusively* for public/open-source code is not actually that useful because most people spend most of their time working on their organization's internal code.

Reviewer: ```suggestion
- Truly valuable code search should cover both public/open-source code and private, proprietary code. The latter is what most developers spend most of their time working on day-to-day.
```

Author: Why do you think this is better?

# Sourcegraph master plan

Everything we do stems from our [purpose](#purpose), to make it so **everyone can code**. We'll accomplish that over the next 30 years by following our strategy:

- 30 years
  - [Purpose](#purpose)
  - [Mission](#mission)
  - [Values](#values)
  - [Big Code](#big-code)
- 3 years
  - [Principles](#principles)
  - [Assumptions](#assumptions)
- 1 year
  - [1-year vision](#1-year-vision)
  - [Product direction](#product-direction)
  - [OKRs](#okrs)

## Purpose

> We want to make it so **everyone can code**. A world where everyone, not just ~0.1% of the world population, can code will see faster and more broadly beneficial technological progress.

Reviewer: I think this document is missing a more coherent bridge between the 1-year vision (make it so existing teams of software professionals can tackle Big Code) and the 30-year purpose (make it so everyone can code). Part of the issue is the rephrasing of the purpose from "everyone builds software" to "everyone can code". The latter has more of an elementary feel ("hey, you wrote 'Hello World' in Java, congrats you can code!") and seems more natural for a company that's tackling the problem of programming education. I believe our goal is not merely to make it so everyone can code, but to foster a world where everyone can *and does* contribute to software used by others. I.e., to make it so everyone can be a producer/contributor to the code economy (alternative phrasing: make it so the code economy is accessible to all). To me, there is a clearer path from "make it so software professions can tackle Big Code" to "make it so everyone can build software / partake in the code economy" than to "make it so everyone can code."

Another thing this document could use is some treatment of the topic of different skill levels / domain expertise in programming itself (and how we help each "level up"). This would help reinforce the point that we are trying to make code accessible to everyone, including groups of different skill levels, spanning from expert engineers to novices to the currently non-technical. But that might be beyond the scope of this PR.

Author: Agree on the missing bridge. I plan to do that later (the PR description mentions this). You ok with merging this now and me adding the other parts later (will work with you on them)?

Author: @beyang ping

Reviewer: Yep, SGTM.

### Background

For thousands of years after writing was invented, most people remained illiterate. Universal literacy seemed unlikely. Is it really possible that every human would be capable of reading and writing? How would literacy benefit the average person? As we now know, every human is capable of and benefits immensely from literacy.

> More recently, around 1976, just 0.2% of the world's population used computers. Two tiny companies sought to make computing universal: Apple's vision then was to create a "bicycle for the mind" in the form of a computer, and Microsoft wanted to put a computer "on every desk and in every home". Though it seemed unlikely at the time, as we now know, everyone is capable of and benefits immensely from having a computer (or a phone).

Reviewer: ```suggestion
More recently, around 1976, just 0.2% of the world's population used computers. Two tiny companies sought to make computing universal: Apple's vision then was to create a "bicycle for the mind" in the form of a computer, and Microsoft wanted to put a computer "on every desk and in every home". Though it seemed unlikely at the time, as we now know, everyone is capable of and benefits immensely from having a computer on their desk, in their home—or in their pocket.
```

> Today, only about 0.1% of the world's population can code. That tiny group has built software that runs the modern world and improves the lives of billions of people. Think of the possibilities if everyone was able to code. All around the world, more people would be able to solve problems and improve their lives by building software. We don't know exactly what these billions of coders will create, but we know that this will bring faster and more broadly beneficial technological progress.

Reviewer: ```suggestion
Today, only about 0.1% of the world's population can code. That tiny group has built software that runs the modern world and improves the lives of billions of people. Think of the possibilities if everyone could code. All around the world, more people would be able to solve problems and improve their lives by building software. We don't know exactly what these billions of coders will create, but we know that this will bring faster and more broadly beneficial technological progress.
```

Our [values](values.md) are the principles we want to uphold as we work toward our mission.

<!-- TODO: these will change, see https://docs.google.com/document/d/1_L-XOT4oPTOuUkT4_JMykneo_wPn2WqC2BXR6_-U5dI/edit -->

- [People](values.md#people)
- [Journey](values.md#journey)
- [Progress](values.md#progress)

## Vision

### 1-year vision {#1-year-vision}

Sourcegraph universal code search is the one place to find and fix things across all code.

- Sourcegraph knows most everything that can be known about your code (from the compiler, build systems, code hosts, runtime metrics, docs, issues, project management, etc.).
- It's easy to find and use this information on Sourcegraph itself and in your editor, code host, and code reviews.
- You use Sourcegraph in your existing workflows for writing, reviewing, understanding, and fixing code.

### 3-year vision {#3-year-vision}

> Sourcegraph understands higher-level software development patterns and manages processes proactively, which lets you spend more time on the higher-leverage tasks that truly require your human brain.

Reviewer 4: The "proactive" bit is still fuzzy to me. Is Sourcegraph being proactive or does it help devs be proactive?

Author: reworded for clarity

- Sourcegraph has purpose-built new workflows that build on these concepts for common tasks. For example, if an application is reported down by your monitoring service, the first step is to visit a Sourcegraph page that shows all relevant information (recent deployments, recent commits, aberrant traces and logs, code owners, etc.) and presents your organization's preferred actions (roll back, create war room chat channel, etc.).
- When you make a change to an API, you can include "metacode" that knows how to update all usages of the changed API. Sourcegraph will show you a compatibility report and propose those changes on all downstream callers along with your change.
- You can discover and more easily use libraries, with statistics about usage patterns and a code marketplace.

> Unlike the [1-year vision](#1-year-vision), our 3-year vision involves Sourcegraph modifying existing workflows and being proactive.

### 10-year vision {#10-year-vision}

> Everyone involved in the software value chain—developers, sales, product, marketing, support, etc.—uses Sourcegraph to coordinate work and share information.

Reviewer 4: I think parens make this more readable compared to dashes.

```suggestion
Everyone involved in the software value chain (e.g., developers, sales, product, marketing, support) uses Sourcegraph to coordinate work and share information.
```
